src/libraries/System.Transactions/src/System/Transactions/Oletx/DtcInterfaces.cs

```diff
@@ -0,0 +1,339 @@
+    InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    internal interface IPhase0EnlistmentShim
+    {
+        void Unenlist();
+
+        void Phase0Done([MarshalAs(UnmanagedType.Bool)] bool voteYes);
+    }
+
+    [Security.SuppressUnmanagedCodeSecurity,
+    ComImport,
+    Guid("5EC35E09-B285-422c-83F5-1372384A42CC"),
+    InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    internal interface IEnlistmentShim
+    {
+        void PrepareRequestDone(OletxPrepareVoteType voteType);
+
+        void CommitRequestDone();
+
+        void AbortRequestDone();
+    }
+
+    [Security.SuppressUnmanagedCodeSecurity,
+    ComImport,
+    Guid("279031AF-B00E-42e6-A617-79747E22DD22"),
+    InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    internal interface ITransactionShim
+    {
+        void Commit();
+
+        void Abort();
+
+        void GetITransactionNative([MarshalAs(UnmanagedType.Interface)] out IDtcTransaction transactionNative);
+
+        void Export(
+            [MarshalAs(UnmanagedType.U4)] uint whereaboutsSize,
+            [MarshalAs(UnmanagedType.LPArray)] byte[] whereabouts,
+            [MarshalAs(UnmanagedType.I4)] out int cookieIndex,
+            [MarshalAs(UnmanagedType.U4)] out uint cookieSize,
+            out CoTaskMemHandle cookieBuffer);
+
+        void CreateVoter(
+            IntPtr managedIdentifier,
+            [MarshalAs(UnmanagedType.Interface)] out IVoterBallotShim voterBallotShim);
+
+        void GetPropagationToken(
+            [MarshalAs(UnmanagedType.U4)] out uint propagationTokeSize,
+            out CoTaskMemHandle propagationToken);
+
+        void Phase0Enlist(
+            IntPtr managedIdentifier,
+            [MarshalAs(UnmanagedType.Interface)] out IPhase0EnlistmentShim phase0EnlistmentShim);
+
+        void GetTransactionDoNotUse(out IntPtr transaction);
+    }
+
+    [Security.SuppressUnmanagedCodeSecurity,
+    ComImport,
+    Guid("27C73B91-99F5-46d5-A247-732A1A16529E"),
+    InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    internal interface IResourceManagerShim
+    {
+        void Enlist(
+            [MarshalAs(UnmanagedType.Interface)] ITransactionShim transactionShim,
+            IntPtr managedIdentifier,
+            [MarshalAs(UnmanagedType.Interface)] out IEnlistmentShim enlistmentShim);
+
+        void Reenlist(
+            [MarshalAs(UnmanagedType.U4)] uint prepareInfoSize,
+            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 1)] byte[] prepareInfo,
+            out OletxTransactionOutcome outcome);
+
+        void ReenlistComplete();
+    }
+
+    [Security.SuppressUnmanagedCodeSecurity,
+    ComImport,
+    Guid("467C8BCB-BDDE-4885-B143-317107468275"),
+    InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    internal interface IDtcProxyShimFactory
+    {
+        // See https://github.com/dotnet/runtime/issues/45633
+        [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(CoTaskMemHandle))]
```

Comment:
It cost me about a day to get to the bottom of this one. Probably better to keep the ctor from getting trimmed via an XML (as in #45911) rather than via an attribute here (in case other path ever needs it).

Reply:
It won't be a concern if `CoTaskMemHandle` has an accessible parameterless constructor and we use the `LibrayImport` source generator; the generated code that will create the handle will be visible to the linker.

Reply:
Since I'm porting the C++ code to C#, CoTaskMemHandle will no longer exist.